backend/mailru/mailru.go

```diff
@@ -0,0 +1,2356 @@
+		res, err := f.srv.CallJSON(&opts, nil, &response)
+		return shouldRetry(res, err, f, &opts)
+	})
+
+	if err == nil && response.Body != "" {
+		return api.PublicLinkURL + response.Body, nil
+	}
+	if err == nil {
+		return "", errors.New("server returned empty link")
+	}
+	if apiErr, ok := err.(*api.FileErrorResponse); ok && apiErr.Status == 404 {
+		return "", fs.ErrorObjectNotFound
+	}
+	return "", err
+}
+
+// CleanUp permanently deletes all trashed files/folders
+func (f *Fs) CleanUp(ctx context.Context) error {
+	fs.Debugf(f, ">>> CleanUp")
+
+	token, err := f.accessToken()
+	if err != nil {
+		return err
+	}
+
+	data := url.Values{
+		"email":   {f.opt.Username},
+		"x-email": {f.opt.Username},
+	}
+	opts := rest.Opts{
+		Method: "POST",
+		Path:   "/api/m1/trashbin/empty",
+		Parameters: url.Values{
+			"access_token": {token},
+		},
+		Body:        strings.NewReader(data.Encode()),
+		ContentType: api.BinContentType,
+	}
+
+	var response api.CleanupResponse
+	err = f.pacer.Call(func() (bool, error) {
+		res, err := f.srv.CallJSON(&opts, nil, &response)
+		return shouldRetry(res, err, f, &opts)
+	})
+	if err != nil {
+		return err
+	}
+
+	switch response.StatusStr {
+	case "200":
+		return nil
+	default:
+		return fmt.Errorf("cleanup failed (%s)", response.StatusStr)
+	}
+}
+
+// About gets quota information
+func (f *Fs) About(ctx context.Context) (*fs.Usage, error) {
+	fs.Debugf(f, ">>> About")
+
+	token, err := f.accessToken()
+	if err != nil {
+		return nil, err
+	}
+	opts := rest.Opts{
+		Method: "GET",
+		Path:   "/api/m1/user",
+		Parameters: url.Values{
+			"access_token": {token},
+		},
+	}
+
+	var info api.UserInfoResponse
+	err = f.pacer.Call(func() (bool, error) {
+		res, err := f.srv.CallJSON(&opts, nil, &info)
+		return shouldRetry(res, err, f, &opts)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	total := info.Body.Cloud.Space.BytesTotal
+	used := int64(info.Body.Cloud.Space.BytesUsed)
+
+	usage := &fs.Usage{
+		Total: fs.NewUsageValue(total),
+		Used:  fs.NewUsageValue(used),
+		Free:  fs.NewUsageValue(total - used),
+	}
+	return usage, nil
+}
+
+// Put the object
+// Copy the reader in to the new object which is returned
+// The new object may have been created if an error is returned
+func (f *Fs) Put(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	o := &Object{
+		fs:      f,
+		remote:  src.Remote(),
+		size:    src.Size(),
+		modTime: src.ModTime(ctx),
+	}
+	fs.Debugf(f, ">>> Put: %q %d '%v'", o.remote, o.size, o.modTime)
+	return o, o.Update(ctx, in, src, options...)
+}
+
+// Update an existing object
+// Copy the reader into the object updating modTime and size
+// The new object may have been created if an error is returned
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
+	wrapIn := in
+	size := src.Size()
+	if size < 0 {
+		return errors.New("mailru does not support streaming uploads")
```

Comment:
This will mean that `rclone mount` uploads have to be spooled to disk first...  It would be good to support if possible, but some backends just can't do it!

Reply:
Unfortunately Mailru wants size and hash in advance. That's the base of their deduplication logic. OTOH they let you skip the whole body if they have one :)